CanonicalABI.md

```diff
@@ -0,0 +1,1082 @@
+      len = read_4_bytes(src, ptr + 4) as wasm_i32
+      lift(direction, src, list<$ty>, [ptr, len])
+    }
+
+    bitflags<$fields> => {
+      bits = 0
+      for i in 0..align_to($fields.len(), 32):
+        bits = (bits << 32) | read_4_bytes(src, ptr + i * 4)
+      bits as bitflags<$fields>
+    }
+
+    record<$fields> => {
+      result = {}
+      for $field in $fields:
+        ptr = align_to(ptr, align(direction, $field))
+        result[$field] = lift_from_memory(direction, src, ptr, $field)
+        ptr += size(direction, $field)
+      result as record<$fields>
+    }
+
+    variant<$cases> => {
+      match val {
+        $case(payload) => {
+          # read the discriminant, using the smallest size available
+          match $cases.len() {
+            0 => unreachable,
+            n if n <= 1<<8 => {
+              discrim = read_1_byte(dst, ptr)
+              ptr += 1
+            }
+            n if n <= 1<<16 => {
+              discrim = read_2_bytes(dst, ptr)
+              ptr += 2
+            }
+            n if n <= 1<<32 => {
+              discrim = read_4_bytes(dst, ptr)
+              ptr += 4
+            }
+            // ...
+          }
+
+          # Recursively read the payload, if present for this case. Note that
+          # we align the pointer to the whole variant's alignment to ensure that
+          # all payloads start at the same address.
+          if payload:
+            ptr = align_to(ptr, align(direction, variant $cases))
+            payload_val = lift_from_memory(direction, dst, ptr, payload)
+
+          (discrim, payload_val) as variant<$cases>
+        }
+      }
+    }
+
+    # defer validation and such to `lift`
+    handle<$resource> => {
+      val = read_4_bytes(src, ptr)
+      lift(direction, src, handle<$resource>, [val as wasm_i32])
+    }
+
+    push-buffer<$ty> | pull-buffer<$ty>(val) => {
+      assert(direction == "import") # see comments above
+      handle = read_4_bytes(src, ptr)
+      offset = read_4_bytes(src, ptr + 4)
+      len = read_4_bytes(src, ptr + 8)
+      lift(direction, src, ty, [handle as wasm_i32, offset as wasm_i32, len as wasm_i32])
+    }
+```
+
+## Helper `TABLE` class
+
+In Python-like pseudo-code this is how the `TABLE` referenced above in
+lifting/lowering will work for managing indices. Note that the implementation
+provided here is not canonical. The indices used to refer to handles and buffers
+can vary between engines and their values should not be relied upon. A sample
+implementation, however, is provided here to show some of the desired semantics.
+
+```
+class TABLE:
+  instances = Map<Instance, Slab<Handle>>
+  buffers = Vec<(Buffer, Instance)>
+
+  # This function will flag that `instance` has access to the `handle` provided.
+  #
+  # Returns a 32-bit integer which can later be passed to `get` to retrieve the
+  # original `handle`. This function always succeeds, unless the table
+  # overflows from too many insertions.
```

Comment:
Thinking more on this and the `realloc`, this raises a few questions for me:

* How come we're trying to future proof against future more general adapters? In that world wouldn't you be able to specify whatever adapter code you wanted and nothing would be baked in? For example this `TABLE` would be entirely implemented in adapter code (or by engines with `resource.new` and such), so if this is purely an implementation of the canonical ABI, and we're assigning just one set of semantics to the canonical ABI, I'm not sure why we'd need to future proof so modules can do something on their own.
* Do your concerns here also apply to `malloc` allocation failure? I've currently documented that `canonical_abi_malloc` is expected to trap if allocation fails, but if we're going the route of future-proofing should it also return -1?

Although I suppose that this could also be interpreted in the context of your other comment where the TABLE bits here should be intrinsics, not implicitly happening as part of the glue. That would mean that modules themselves would receive the -1 and decide what to do, which in that case I would agree is not future proofing it's just letting modules be more flexible.

(my question about future-proofing though would still apply to malloc/realloc where those are still only called by adapters of the canonical abi)

Reply:
The interaction I think we have to future-proof for is when a custom-adapted component calls a canonically-adapted component.  If the canonically-adapted component effectively requires the list size to be known up-front (so it can issue a single `malloc` call), but the custom adapter is doing, say, UTF-16 decoding (where the number-of-USVs isn't known until the end), then that forces the glue to buffer up the decoded `char`s (an intermediate copy) in a first USV-counting pass, then `malloc`, then copy the `char`s into the malloced buffer.

Incidentally, I think we need not just `realloc` to be defined, but to actually used in practice (in place of `malloc`).  I still worry we'll get components with broken `realloc`s that only show up with custom adapters...